feat: add vector settings from Python SDK

frascuchon opened the pull request:
<!-- Thanks for your contribution! As part of our Community Growers initiative 🌱, we're donating Justdiggit bunds in your name to reforest sub-Saharan Africa. To claim your Community Growers certificate, please contact David Berenstein in our Slack community or fill in this form https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

This PR includes the base definition for working with vector settings from the Python SDK and also implements the `add_vector_settings` and `vector_settings_by_name` and list vector settings for both, local and remote datasets.

Refs #3028 

**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactor (change restructuring the codebase without changing functionality)
- [X] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And ideally, reference `tests`)

Local test workflow

**Checklist**

- [ ] I added relevant documentation
- [ ] I followed the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See https://keepachangelog.com/)

frascuchon committed:
chore: Add new methods to the abstract class

frascuchon committed:
feat: Adding vector settings for both, local aand remote datasets

frascuchon committed:
feat: Adding API calls for adding and listing vector settings

frascuchon committed:
tests: Include basic integration tests

frascuchon committed:
chore: Adding changelog

frascuchon committed:
feat: Adding vector_settings_by_name

frascuchon committed:
tests: Update tests

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon committed:
chore: Adding missing methods to the test dataset class

frascuchon committed:
Merge branch 'feat/add-vector-settings-from-python-sdk' of github.com:argilla-io/argilla into feat/add-vector-settings-from-python-sdk

gabrielmbmb commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -453,6 +466,21 @@ def delete_metadata_properties(self, *args, **kwargs):
 
+    @abstractmethod
+    def add_vector_settings(self, *args, **kwargs):
+        """Adds a new `vecto_settings` to the current `FeedbackDataset`."""
```
```suggestion
        """Adds a new `vector_settings` to the current `FeedbackDataset`."""
```

gabrielmbmb commented on src/argilla/client/feedback/dataset/local/dataset.py:
```diff
@@ -247,6 +254,26 @@ def add_metadata_property(
+    def vector_settings_by_name(self, name: str) -> VectorSettings:
+        vector_settings = self._vector_settings.get(name)
+        if not vector_settings:
+            raise KeyError(f"Vector settings with name {name!r} does not exist in the dataset.")
```
```suggestion
            raise KeyError(f"Vector settings with name '{name!r}' does not exist in the dataset.")
```

gabrielmbmb commented on src/argilla/client/feedback/dataset/remote/dataset.py:
```diff
@@ -591,6 +600,34 @@ def add_metadata_property(
+                dimensions=vector_settings.dimensions,
+            ).parsed
+        except AlreadyExistsApiError:
+            raise ValueError(f"Vector settings with name {vector_settings.name!r} already exists")
```
```suggestion
            raise ValueError(f"Vector settings with name '{vector_settings.name!r}' already exists.")
```

gabrielmbmb replied:
shouldn't we try to catch all the exceptions here?

gabrielmbmb committed:
Apply suggestions from code review

frascuchon committed:
chore: Fix error message detail

frascuchon committed:
tests: Adding basic unit tests

jfcalvo commented on src/argilla/client/feedback/dataset/local/dataset.py:
```diff
@@ -154,6 +155,12 @@ def __init__(
         )
 
         self._records = []
+        self._vector_settings: Dict[str, VectorSettings] = {}
```
Why are we using a dictionary instead of a list of `VectorSettings`? Actually I can see in the Notion documentation that we are using a list:

```python
dataset = rg.FeedbackDataset(
	fields=...,
	questions=...,
	✨ vectors_settings=[VectorSettings("name", dim=384, description="...")], # all vectors are optional
	guidelines=...,
)
```

frascuchon replied:
The dictionary is used for better handling the `vector_settings_by_name` and the `add_vector_settings` validations. Using a dict internally simplifies these methods.

jfcalvo commented on src/argilla/client/feedback/dataset/local/dataset.py:
```diff
@@ -247,6 +254,26 @@ def add_metadata_property(
+        self._vector_settings[vector_settings.name] = vector_settings
+        return vector_settings
+
+    def update_vector_settings(self, *args, **kwargs):
+        pass
+
+    def delete_vector_settings(self, *args, **kwargs):
+        pass
```
I understand that these methods are not implemented because they doesn't have any sense in the local `FeedbackDataset`, or do we have another reason for it?

jfcalvo commented on CHANGELOG.md:
```diff
@@ -22,6 +22,7 @@ These are the section headers that we use:
 - Added `GET /api/v1/datasets/:dataset_id/vectors-settings` endpoint for listing the vectors settings for a dataset. ([#3776](https://github.com/argilla-io/argilla/pull/3776))
 - Added `DELETE /api/v1/vectors-settings/:vector_settings_id` endpoint for deleting a vector settings. ([#3776](https://github.com/argilla-io/argilla/pull/3776))
 - Added `GET /api/v1/records/:record_id` endpoint to get a specific record. ([#4039](https://github.com/argilla-io/argilla/pull/4039))
+- Added support for adding and getting vector settings from Python SDK ([#4055](https://github.com/argilla-io/argilla/pull/4055))
```
I would add a line for every one of the added methods.

Something similar to:

```suggestion
- Added `add_vector_settings` function to local and remote FeedbackDataset. ([#4055](https://github.com/argilla-io/argilla/pull/4055))
...
```

jfcalvo commented on src/argilla/client/feedback/dataset/remote/dataset.py:
```diff
@@ -550,6 +553,12 @@ def metadata_properties(self) -> List["AllowedRemoteMetadataPropertyTypes"]:
+    def vector_settings(self) -> List[RemoteVectorSettings]:
+        """Retrieves the `vector_settings` of the current dataset from Argilla"""
+        response_model = datasets_api_v1.list_vector_settings(client=self._client, id=self.id).parsed
+        return [RemoteVectorSettings.from_orm(vector_settings) for vector_settings in response_model.items]
```
Maybe use `response` instead?

```suggestion
        response = datasets_api_v1.list_vector_settings(client=self._client, id=self.id)
        
        return [RemoteVectorSettings.from_orm(vector_settings) for vector_settings in response.parsed.items]
```

jfcalvo commented on src/argilla/client/feedback/dataset/remote/dataset.py:
```diff
@@ -550,6 +553,12 @@ def metadata_properties(self) -> List["AllowedRemoteMetadataPropertyTypes"]:
+    @property
+    def vector_settings(self) -> List[RemoteVectorSettings]:
+        """Retrieves the `vector_settings` of the current dataset from Argilla"""
+        response_model = datasets_api_v1.list_vector_settings(client=self._client, id=self.id).parsed
```
For other similar methods we are capturing every exception and raising a `RuntimeError`. Do we want to the same here?

jfcalvo commented on src/argilla/client/feedback/dataset/remote/dataset.py:
```diff
@@ -591,6 +600,34 @@ def add_metadata_property(
+
+        return RemoteVectorSettings.from_orm(new_vector_settings)
+
+    def update_vector_settings(self, *args, **kwargs):
+        pass
+
+    def delete_vector_settings(self, *args, **kwargs):
+        pass
```
This is here because we want to do it later?

frascuchon replied:
yes

jfcalvo commented on tests/unit/client/feedback/dataset/remote/test_dataset.py:
```diff
@@ -233,3 +238,39 @@ def test_push_to_huggingface_warnings(
+            vector_settings=VectorSettings(name=expected_name, dimensions=expected_dimensions)
+        )
+
+        mock_httpx_client.post.assert_called_with(
```
```suggestion
        mock_httpx_client.post.assert_called_once_with(
```

frascuchon committed:
Apply suggestions from code review

Co-authored-by: José Francisco Calvo <jose@argilla.io>

frascuchon committed:
Merge branch 'feature/feedback-dataset-semantic-similarity' into feat/add-vector-settings-from-python-sdk

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon committed:
feat: Add vector settings on FeedbackDataset creation

frascuchon committed:
chore: Adding CHANGELOG

frascuchon committed:
fix: Avoid remove API error details

frascuchon committed:
fix: Remove create dataset on vector settings creation error

frascuchon committed:
tests: Add missing test

frascuchon committed:
chore: Add missing imports

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon committed:
fix: Parsing optional params

jfcalvo commented on CHANGELOG.md:
```diff
@@ -25,13 +25,18 @@ These are the section headers that we use:
 - Added support to include vectors for `GET /api/v1/me/datasets/:dataset_id/records` endpoint response using `include` query param. ([#4063](https://github.com/argilla-io/argilla/pull/4063))
 - Added support to include vectors for `POST /api/v1/me/datasets/:dataset_id/records/search` endpoint response using `include` query param. ([#4063](https://github.com/argilla-io/argilla/pull/4063))
+- Added `add_vector_settings` method to local and remote FeedbackDataset. ([#4055](https://github.com/argilla-io/argilla/pull/4055))
+- Added `vector_settings_by_name` method to local and remote Feedback datasets. ([#4055](https://github.com/argilla-io/argilla/pull/4055))
```
```suggestion
- Added `add_vector_settings` method to local and remote `FeedbackDataset`. ([#4055](https://github.com/argilla-io/argilla/pull/4055))
- Added `vector_settings_by_name` method to local and remote `FeedbackDataset`. ([#4055](https://github.com/argilla-io/argilla/pull/4055))
```

jfcalvo commented on CHANGELOG.md:
```diff
@@ -25,13 +25,18 @@ These are the section headers that we use:
 - Update `PATCH /api/v1/datasets/:dataset_id/records` endpoint to allow to update records with `vectors`. ([#4062](https://github.com/argilla-io/argilla/pull/4062))
 - Update `PATCH /api/v1/records/:record_id` endpoint to allow to update record with `vectors`. ([#4062](https://github.com/argilla-io/argilla/pull/4062))
 - Update `BaseElasticAndOpenSearchEngine.index_records` method to also index record vectors. ([#4062](https://github.com/argilla-io/argilla/pull/4062))
+- Update `FeedbackDataset.__init__` to allow passing a vector_settings list. ([#4055](https://github.com/argilla-io/argilla/pull/4055))
```
```suggestion
- Update `FeedbackDataset.__init__` to allow passing a list of vector settings. ([#4055](https://github.com/argilla-io/argilla/pull/4055))
```

jfcalvo commented on src/argilla/client/feedback/dataset/local/dataset.py:
```diff
@@ -247,6 +259,26 @@ def add_metadata_property(
         return metadata_property
 
+    def vector_settings_by_name(self, name: str) -> VectorSettings:
+        vector_settings = self._vector_settings.get(name)
+        if not vector_settings:
+            raise KeyError(f"Vector settings with name '{name!r}' does not exist in the dataset.")
+
+        return vector_settings
```
Not sure if raising an exception here is too radical. Maybe we should return `None` instead?

```suggestion
        return self._vector_settings.get(name)
```

frascuchon replied:
This is aligned to other `*_by_name` methods that we have but could make sense to change the signature to simplify conditional creation workflows (if this setting is not created, then create it, instead of try-catch and create on catch block)

I've created an issue for this. See https://github.com/argilla-io/argilla/issues/4101

jfcalvo commented on src/argilla/client/feedback/dataset/local/dataset.py:
```diff
@@ -247,6 +259,26 @@ def add_metadata_property(
+
+        return vector_settings
+
+    def add_vector_settings(self, vector_settings: VectorSettings) -> VectorSettings:
+        if self._vector_settings.get(vector_settings.name):
+            raise ValueError(f"Vector settings with name '{vector_settings.name}' already exists in the dataset.")
+
+        self._vector_settings[vector_settings.name] = vector_settings
+        return vector_settings
```
Something similar here. Do we want to raise an exception or silently overwrite the specified vector settings instead?
```suggestion
    def add_vector_settings(self, vector_settings: VectorSettings) -> VectorSettings:
        self._vector_settings[vector_settings.name] = vector_settings
        
        return vector_settings
```

frascuchon replied:
Again, this is aligned with other `add_*` methods. Anyway, in this case could make sense to warn at least the user since there is already a previous vector setting, and can be useful to users to detect wrong configuration definitions.

gabrielmbmb commented on src/argilla/client/feedback/dataset/remote/dataset.py:
```diff
@@ -550,6 +553,13 @@ def metadata_properties(self) -> List["AllowedRemoteMetadataPropertyTypes"]:
+        """Retrieves the `vector_settings` of the current dataset from Argilla"""
+        response = datasets_api_v1.list_vector_settings(client=self._client, id=self.id)
+
+        return [RemoteVectorSettings.from_orm(vector_settings) for vector_settings in response.parsed.items]
```
In the rest of the remote schemas we're defining the `from_api` method instead of using the `from_orm` method, should we do the same here?

frascuchon replied:
Yes, I will use the same mechanism that we're using for the rest of the schemas. Anyway, maybe make sense to use the `from_orm` + [pydantic data bindings](https://docs.pydantic.dev/1.10/usage/models/#data-binding) which are commonly used for this kind of mapping. What do you think?

frascuchon replied:
Done

jfcalvo commented on src/argilla/client/feedback/dataset/remote/dataset.py:
```diff
@@ -591,6 +601,34 @@ def add_metadata_property(
 
         return ArgillaMixin._parse_to_remote_metadata_property(metadata_property=metadata_property, client=self._client)
 
+    def vector_settings_by_name(self, name: str) -> RemoteVectorSettings:
+        # TODO: Maybe make sense to have a query param in api.list_vector_settings to filter by name
+        for vector_settings in self.vector_settings:
+            if vector_settings.name == name:
+                return vector_settings
+
+        raise KeyError(f"Vector settings with name {name!r} does not exist in the dataset.")
```
If we choose to go for not raising an error when the vector_settings is not found then maybe we should do something similar here.

frascuchon replied:
I give you here the same argument as before. See https://github.com/argilla-io/argilla/issues/4101

jfcalvo commented on src/argilla/client/feedback/schemas/vector_settings.py:
```diff
@@ -0,0 +1,24 @@
+    name: str
+    dimensions: int
+    # TODO Uncomment when is supported
+    # title: Optional[str] = None
```
```suggestion
class VectorSettings(BaseModel):
    name: str
    # TODO: Uncomment when is supported
    # title: Optional[str] = None
    dimensions: int
```

jfcalvo commented on src/argilla/client/sdk/v1/datasets/api.py:
```diff
@@ -471,6 +473,45 @@ def add_metadata_property(
+        response_obj = Response.from_httpx_response(response)
+        response_obj.parsed = FeedbackListVectorSettingsModel(**response.json())
+        return response_obj
+    return handle_response_error(response)
```
```suggestion
    url = f"/api/v1/datasets/{id}/vectors-settings"

    response = client.get(url=url)
    if response.status_code == 200:
        response_obj = Response.from_httpx_response(response)
        response_obj.parsed = FeedbackListVectorSettingsModel(**response.json())
        return response_obj
    
    return handle_response_error(response)
```

jfcalvo commented on src/argilla/client/sdk/v1/datasets/models.py:
```diff
@@ -128,5 +128,17 @@ class FeedbackRecordsMetricsModel(BaseModel):
+    name: str
+    dimensions: int
+    inserted_at: datetime
+    updated_at: datetime
```
```suggestion
class FeedbackVectorSettingsModel(BaseModel):
    id: UUID
    name: str
    # TODO: Uncomment when is supported
    # title: Optional[str] = None
    dimensions: int
    inserted_at: datetime
    updated_at: datetime
```

frascuchon committed:
Apply suggestions from code review

Co-authored-by: José Francisco Calvo <jose@argilla.io>

frascuchon committed:
refactor: Using .from_api method for RemoteVectorSettings

frascuchon committed:
Update src/argilla/client/feedback/schemas/vector_settings.py

Co-authored-by: José Francisco Calvo <jose@argilla.io>

frascuchon committed:
Update src/argilla/client/sdk/v1/datasets/models.py

Co-authored-by: José Francisco Calvo <jose@argilla.io>

frascuchon committed:
Update src/argilla/client/sdk/v1/datasets/api.py

Co-authored-by: José Francisco Calvo <jose@argilla.io>

frascuchon committed:
Merge branch 'feature/feedback-dataset-semantic-similarity' into feat/add-vector-settings-from-python-sdk

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

jfcalvo commented on src/argilla/client/feedback/dataset/local/mixins.py:
```diff
@@ -445,6 +449,26 @@ def list(cls: Type["FeedbackDataset"], workspace: Optional[str] = None) -> List[
+                        client=client,
+                        id=id,
+                        title=vector_settings.name,
+                        name=vector_settings.name,
```
```suggestion
                        name=vector_settings.name,
                        title=vector_settings.name,
                        
```

frascuchon committed:
Update src/argilla/client/feedback/dataset/local/mixins.py

Co-authored-by: José Francisco Calvo <jose@argilla.io>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon committed:
feat: Support for vectors creation when adding  records from Python SDK (#4043)

<!-- Thanks for your contribution! As part of our Community Growers
initiative 🌱, we're donating Justdiggit bunds in your name to reforest
sub-Saharan Africa. To claim your Community Growers certificate, please
contact David Berenstein in our Slack community or fill in this form
https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

This PR allows Python SDK users to provide vectors when adding records
to a feedback dataset. Also, the default error handler mechanism has
been fixed to include more info on API errors

This [PR](https://github.com/argilla-io/argilla/pull/4055) must be
merged first

Refs [#3028](https://github.com/argilla-io/argilla/issues/3028)

**Type of change**

(Please delete options that are not relevant. Remember to title the PR
according to the type of change)

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactor (change restructuring the codebase without changing
functionality)
- [X] Improvement (change adding some improvement to an existing
functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And
ideally, reference `tests`)

- [ ] Test A
- [ ] Test B

**Checklist**

- [ ] I added relevant documentation
- [ ] I followed the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

---------

Co-authored-by: José Francisco Calvo <josefranciscocalvo@gmail.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: José Francisco Calvo <jose@argilla.io>

frascuchon committed:
Merge branch 'feature/feedback-dataset-semantic-similarity' into feat/add-vector-settings-from-python-sdk

frascuchon committed:
fix CHANGELOG.md versions
